src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,491 @@
+                var sb = new System.Text.StringBuilder(4096);
+                while(EnsureData())
+                {
+                    sb.Append(charBuffer, charRead, charWrite-charRead);
+                    charRead = charWrite;
+                }
+                return sb.ToString();
+            }
+
+            public async override Task<string> ReadToEndAsync()
+            {
+                var sb = new System.Text.StringBuilder(4096);
+                while(await EnsureDataAsync().ConfigureAwait(false))
+                {
+                    sb.Append(charBuffer, charRead, charWrite-charRead);
+                    charRead = charWrite;
+                }
+                return sb.ToString();
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                stream?.Dispose();
+            }
+
+            bool EnsureData()
+            {
+                if (charRead == charWrite)
+                {
+                    try { return EnsureDataAsync().GetAwaiter().GetResult(); }
+                    catch (OperationCanceledException) { return false; } // non-async callers aren't expecting OperationCanceledException
+                }
+                return true;
+            }
+
+            async Task<bool> EnsureDataAsync()
+            {
+                if (charRead == charWrite)
+                {
+                    CancelToken.ThrowIfCancellationRequested();
+                    charRead = charWrite = 0;
+                    do
+                    {
+                        int bytesRead = await stream.ReadAsync(byteBuffer, 0, byteBuffer.Length, CancelToken).ConfigureAwait(false);
+                        if (bytesRead == 0) return false;
+                        charWrite = decoder.GetChars(byteBuffer, 0, bytesRead, charBuffer, 0);
+                    } while(charWrite == 0);
+                }
+                return true;
+            }
+
+            int Read(char[] buffer, int index, int count, bool block)
+            {
+                int start = index;
+                while(count != 0 && EnsureData())
+                {
+                    int toRead = Math.Min(count, charWrite-charRead);
+                    Array.Copy(charBuffer, charRead, buffer, index, toRead);
+                    charRead += toRead;
+                    index += toRead;
+                    if (!block) break;
+                    count -= toRead;
+                }
+                return index - start;
+            }
+
+#if NETCOREAPP2_1
+            async ValueTask<int> ReadAsync(Memory<char> buffer, bool block)
+            {
+                int totalRead = 0;
+                while(buffer.Length != 0 && (charRead != charWrite || await EnsureDataAsync().ConfigureAwait(false)))
+                {
+                    int toRead = Math.Min(buffer.Length, charWrite-charRead);
+                    charBuffer.AsMemory(0, toRead).CopyTo(buffer);
```

Comment:
I think `charRead` must be given for `start`?
```csharp
charBuffer.AsMemory(charRead, toRead).CopyTo(buffer);
```
Also, I believe `AsSpan` be used instead as the span would remain on the stack.

Reply:
Fixed.